### ISSUE TITLE
Cannot read property 'animatableElement' of undefined - aurelia-templating-router 1.0.1 regression

### ISSUE THREAD

**Issue by brenthoneybone** [aurelia/templating#518]
**I'm submitting a bug report**

* **Library Version:**
Not sure how to find this when using the Aurelia CLI installation, here's the installed tree from NPM for Aurelia Libraries:
├─┬ aurelia-animator-css@1.0.1
│ ├── aurelia-metadata@1.0.3
│ ├── aurelia-pal@1.2.0
│ └─┬ aurelia-templating@1.2.0
│   ├── aurelia-binding@1.1.1
│   ├── aurelia-loader@1.0.0
│   ├── aurelia-path@1.1.1
│   └── aurelia-task-queue@1.1.0
├─┬ aurelia-bootstrapper@1.0.1
│ ├── aurelia-event-aggregator@1.0.1
│ ├── aurelia-framework@1.0.8
│ ├── aurelia-history@1.0.0
│ ├── aurelia-history-browser@1.0.0
│ ├── aurelia-loader-default@1.0.0
│ ├── aurelia-logging-console@1.0.0
│ ├── aurelia-pal-browser@1.1.0
│ ├── aurelia-polyfills@1.1.1
│ ├─┬ aurelia-router@1.1.1
│ │ └── aurelia-route-recognizer@1.1.0
│ ├── aurelia-templating-binding@1.2.0
│ ├── aurelia-templating-resources@1.2.0
│ └── aurelia-templating-router@1.0.1
├─┬ aurelia-cli@0.23.0
├─┬ aurelia-testing@1.0.0-beta.2.0.1
│ └── aurelia-logging@1.2.0
├─┬ aurelia-tools@0.2.4

**Please tell us about your environment:**
* **Operating System:**
Ubuntu 16.04

* **Node Version:**
7.3.0

* **NPM Version:**
3.10.10

* **JSPM OR Webpack AND Version**
Aurelia CLI v0.23.0

* **Browser:**
Chrome 55.0.2883.75 (64-bit), Chrome 56.0.2924.21 beta (64-bit) | Firefox 50.1.0

* **Language:**
ESNext

**Current behavior:**
This is in addition to issue #470, I believe it's a separate issue from the bug that @jdanyow has already addressed on that issue. Please refer to my posts on that issue for any additional information.

On page load I receive the following error: `ERROR [app-router] TypeError: Cannot read property 'animatableElement' of undefined`

This seems to be directly tied with `swap-order="with"` and seems to be a regression in aurelia-templating-router 1.0.1 from 1.0.0. Reverting aurelia-templating-router back to 1.0.0 fixes the problem.

After a little digging, the error stems from:
SwapStrategies.with which calls - 

```javascript
return Promise.all([viewSlot.remove(previousView, true), promise]);
```

The ViewSlot remove method then looks for that `previousView` object within its children, however, the children array is empty resulting in the following method call of `ViewSlot.removeAt()` to have an invalid index which means the look up for the view again fails 

```javascript
let view = this.children[index]; // undefined.
```

Clicking between the two pages results in the same error being repeated and the new view to be attached, however the old view still remains and starts stacking down the page.

I've created a Gist, however it seems to 404 when loading any additional resources but the error still occurs
https://gist.run/?id=c75826dc0c8e0006b7a42d575f1392a9

I have also done a local bare bones install to replicate the problem which can be found here:
[aurelia-bug-replication.zip](https://github.com/aurelia/templating/files/707063/aurelia-bug-replication.zip)

**Comment by thomas-darling** [comment]
Yep, we're also seeing this, and this time it's critical, as we can't just downgrade, due to other issues.
Right now we're temporarily dodging the issue with this hack, which is far from ideal:

```ts
import {RouterView} from "aurelia-templating-router";

/*tslint:disable: only-arrow-functions no-invalid-this */
const swapFunc = RouterView.prototype.swap;
RouterView.prototype.swap = function (...args)
{
    if (this.swapOrder === "with")
    {
        this.swapOrder = undefined;
    }

    swapFunc.apply(this, args);
};
/*tslint:enable*/
```

**Comment by EisenbergEffect** [comment]
I think this is fixed in master already. It's just pending the next release, which is hopefully going out next week. That's the plan at least.

**Comment by bbarne8** [comment]
I'm not sure if this has been fixed yet. I updated and added all of the changes to master since the release and I was still having the issue. I think I may have found the problem though.

This issue appears to be in the swap function in router-view.js of the aurelia-templating-router. 

Currently, previousView is defined inside of the work function definition like this:
```javascript

//router-view.js

  swap(viewPortInstruction) {
    let layoutInstruction = viewPortInstruction.layoutInstruction;

    let work = () => {
//////////////////////////////////////////////////////////////////////////
     let previousView = this.view; ////This is not being correctly set 
//////////////////////////////////////////////////////////////////////////
      let swapStrategy;
      let viewSlot = this.viewSlot;

      swapStrategy = this.swapOrder in swapStrategies
                  ? swapStrategies[this.swapOrder]
                  : swapStrategies.after;

      swapStrategy(viewSlot, previousView, () => {
        return Promise.resolve().then(() => {
          return viewSlot.add(this.view);
        }).then(() => {
          this._notify();
        });
      });
    };
...    
```

As it stands, previousView is not being set with the correct object for the Previous View and is failing to be found as it should when it is being removed.

When I changed previousView to be defined in the scope of the swap function everything seemed to work as it should and it passes all tests.

It appears that all that needs to be done is simply to move the previousView definition.

```javascript
//router-view.js

  swap(viewPortInstruction) {
    let layoutInstruction = viewPortInstruction.layoutInstruction;
//////////////////////////////////////////////////////////////////////////
    let previousView = this.view;  ////This is now being correctly set 
//////////////////////////////////////////////////////////////////////////

    let work = () => {
   
      let swapStrategy;
      let viewSlot = this.viewSlot;

      swapStrategy = this.swapOrder in swapStrategies
                  ? swapStrategies[this.swapOrder]
                  : swapStrategies.after;

      swapStrategy(viewSlot, previousView, () => {
        return Promise.resolve().then(() => {
          return viewSlot.add(this.view);
        }).then(() => {
          this._notify();
        });
      });
    };
...    
```

You can easily reproduce the error here:
https://github.com/bbarne8/AureliaSwapOrderRepro

**Comment by mroseboom** [comment]
This issue could be closed, because this is included within the merged pr [aurelia/templating-router#55](https://github.com/aurelia/templating-router/commit/959855ad660da1fb351199982b07c17016bdfbeb)
